pkg/protocol/engine/accounts/accountsledger/manager.go

```diff
@@ -294,11 +311,29 @@ func (m *Manager) rollbackAccountTo(accountData *accounts.AccountData, targetInd
 		accountData.AddPublicKeys(diffChange.PubKeysRemoved...)
 		accountData.RemovePublicKeys(diffChange.PubKeysAdded...)
 
-		// TODO: add safemath package, check for overflows in testcases
-		accountData.ValidatorStake = iotago.BaseToken(int64(accountData.ValidatorStake) - diffChange.ValidatorStakeChange)
-		accountData.DelegationStake = iotago.BaseToken(int64(accountData.DelegationStake) - diffChange.DelegationStakeChange)
-		accountData.StakeEndEpoch = iotago.EpochIndex(int64(accountData.StakeEndEpoch) - diffChange.StakeEndEpochChange)
-		accountData.FixedCost = iotago.Mana(int64(accountData.FixedCost) - diffChange.FixedCostChange)
+		validatorStake, err := safemath.SafeSub(int64(accountData.ValidatorStake), diffChange.ValidatorStakeChange)
+		if err != nil {
+			return false, ierrors.Wrapf(err, "can't retrieve account, validator stake underflow for account (%s) in slot (%d): %d - %d", accountData.ID, diffIndex, accountData.ValidatorStake, diffChange.ValidatorStakeChange)
+		}
+		accountData.ValidatorStake = iotago.BaseToken(validatorStake)
+
+		delegationStake, err := safemath.SafeSub(int64(accountData.DelegationStake), diffChange.DelegationStakeChange)
+		if err != nil {
+			return false, ierrors.Wrapf(err, "can't retrieve account, delegation stake underflow for account (%s) in slot (%d): %d - %d", accountData.ID, diffIndex, accountData.DelegationStake, diffChange.DelegationStakeChange)
+		}
+		accountData.DelegationStake = iotago.BaseToken(delegationStake)
+
+		stakeEpochEnd, err := safemath.SafeSub(int64(accountData.StakeEndEpoch), diffChange.StakeEndEpochChange)
+		if err != nil {
+			return false, ierrors.Wrapf(err, "can't retrieve account, stake end epoch underflow for account (%s) in slot (%d): %d - %d", accountData.ID, diffIndex, accountData.StakeEndEpoch, diffChange.StakeEndEpochChange)
+		}
+		accountData.StakeEndEpoch = iotago.EpochIndex(stakeEpochEnd)
+
+		fixedCost, err := safemath.SafeSub(int64(accountData.FixedCost), diffChange.FixedCostChange)
+		if err != nil {
+			return false, ierrors.Wrapf(err, "can't retrieve account, fixed cost underflow for account (%s) in slot (%d): %d - %d", accountData.ID, diffIndex, accountData.FixedCost, diffChange.FixedCostChange)
+		}
+		accountData.FixedCost = iotago.Mana(fixedCost)
```

Comment:
same as above for all these. int64 won't underflow, it will just return a negative result.